packages/flutter/lib/src/painting/_network_image_io.dart

```diff
@@ -57,63 +58,121 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
-            expectedTotalBytes: total,
-          ));
-        },
-      );
-      if (bytes.lengthInBytes == 0)
+
+      // This will keep reference to [debugNetworkImageHttpClientProvider] tree-shaken
+      // out of release build.
+      HttpClientProvider httpClientProvider;
+      assert(() { httpClientProvider = debugNetworkImageHttpClientProvider; return true; }());
+
+      final _DownloadRequest downloadRequest = _DownloadRequest(port.sendPort,
+          resolved, headers, httpClientProvider);
+      if (_requestPort != null) {
+        // If worker isolate is properly set up([_requestPort] is holding
+        // initialized [SendPort], then just send download request down to it.
+        _requestPort.send(downloadRequest);
+      } else {
+        if (_pendingLoader == null) {
+          // If worker isolate creation was not started, start creation now.
+          _pendingLoadRequests = <_PendingLoadRequest>[];
+
+          _pendingLoader = _setupIsolate()..then((Isolate isolate) {
+              isolate.addErrorListener(RawReceivePort(
+                  (List<dynamic> errorAndStacktrace) {
+                    _cleanupDueToError(errorAndStacktrace[0]);
+                  }).sendPort);
+              isolate.resume(isolate.pauseCapability);
+            }).catchError((dynamic error, StackTrace stackTrace) {
+              _cleanupDueToError(error);
+            });
+        }
+        // Record donwload request so it can either send a request when isolate is ready or handle errors.
+        _pendingLoadRequests.add(_PendingLoadRequest(
+            (SendPort sendPort) { sendPort.send(downloadRequest); },
+            (dynamic error) { downloadRequest.sendPort.send(error); }
+        ));
+      }
+
+      final TransferableTypedData transferable = await completer.future;
+      port.close();
+
+      final Uint8List bytes = transferable.materialize().asUint8List();
+      if (bytes.isEmpty)
         throw Exception('NetworkImage is an empty file: $resolved');
 
-      return PaintingBinding.instance.instantiateImageCodec(bytes);
+      return await PaintingBinding.instance.instantiateImageCodec(bytes);
     } finally {
       chunkEvents.close();
     }
   }
 
+  void _cleanupDueToError(dynamic error) {
+    for (_PendingLoadRequest request in _pendingLoadRequests) {
+      request.handleError(error);
+    }
+    _pendingLoadRequests = null;
+    _pendingLoader = null;
+  }
+
+
+  Future<Isolate> _setupIsolate() {
+    // This is used to get _requestPort [SendPort] that can be used to
+    // communicate with worker isolate: when isolate is spawned it will send
+    // it's [SendPort] over via this [RawReceivePort].
+    // Received [sendPort] can also be [null], which indicates that worker
+    // isolate exited after being idle.
+    final RawReceivePort receivePort = RawReceivePort((SendPort sendPort) {
+      _requestPort = sendPort;
+      if (sendPort == null) {
+        _pendingLoader = null;
```

Comment:
assert that ` _pendingLoadRequests.isEmpty` and return?

Reply:
Done